smalltalksrc/Slang/CCodeGenerator.class.st

```diff
@@ -2253,6 +2261,67 @@ CCodeGenerator >> generateCASTTimes: tast [
 	^ self generateCASTBinary: tast operator: #*
 ]
 
+{ #category : #'CAST translation' }
+CCodeGenerator >> generateCASTTimesRepeat: tast [
+
+	| repeat |
+	repeat := CCompoundStatementNode statements: OrderedCollection new.
+	repeat needsBrackets: false.
+	repeat needsCarriageReturn: false.
+	repeat needsSeparator: false.
+
+	tast receiver value timesRepeat: [ | bloc |
+		bloc := (tast arguments first asCASTIn: self).
+		bloc needsCarriageReturn: false.
+		repeat add: bloc ].
+
+	^ repeat
+]
+
+{ #category : #'CAST translation' }
+CCodeGenerator >> generateCASTToByDoUntil: limitExpr iterativeVariable: itVar negative: negative [
+
+	| operator limit |
+	(limitExpr isSend and: [ 
+		 limitExpr selector = #- and: [ 
+			 limitExpr arguments first isConstant and: [ 
+				 limitExpr arguments first value = 1 ] ] ])
+		ifTrue: [ 
+			operator := negative
+				            ifTrue: [ #> ]
+				            ifFalse: [ #< ].
+			limit := limitExpr receiver ]
+		ifFalse: [ 
+			operator := negative
+				            ifTrue: [ #>= ]
+				            ifFalse: [ #<= ].
+			limit := limitExpr ].
+	^ CBinaryOperatorNode
+		  operator: operator
+		  left: (CIdentifierNode name: itVar)
+		  right: (limit asCASTExpressionIn: self)
+]
+
+{ #category : #'CAST translation' }
+CCodeGenerator >> generateCASTToDo: tast [
+
+	| iterationVar init until step |
+	(tast args last args size = 1) ifFalse: [
+		self error: 'wrong number of block arguments'.
+	].
+	iterationVar := tast arguments last args first.
+	init := CAssignmentNode
+		        lvalue: (CIdentifierNode name: iterationVar)
+		        rvalue: (tast receiver asCASTExpressionIn: self).
+	until := self generateCASTToByDoUntil: tast arguments first iterativeVariable: iterationVar negative: false.
+	step := CIncrementNode object: (CIdentifierNode name: iterationVar).
+	^ CForStatementNode
+		  init: init
+		  until: until
+		  step: step
```

Comment:
Do we have a test for a different step?

Reply:
to:do: is always with incrementing step, I will test it with a different step for to:by:do.